src/System.Data.SqlClient/src/System/Data/Sql/SqlNorm.cs

```diff
@@ -336,51 +326,68 @@ internal sealed class ShortNormalizer : Normalizer
     {
         internal override void Normalize(FieldInfo fi, object obj, Stream s)
         {
-            byte[] b = BitConverter.GetBytes((short)GetValue(fi, obj));
+            short value = (short)GetValue(fi, obj);
+            Span<byte> b = stackalloc byte[2];
+
             if (!_skipNormalize)
             {
-                Array.Reverse(b);
+                BinaryPrimitives.WriteInt16BigEndian(b, value);
                 b[0] ^= 0x80;
             }
-            s.Write(b, 0, b.Length);
+            else
+            {
+                BinaryPrimitives.WriteInt16LittleEndian(b, value);
```

Comment:
Do you know why the original logic was reversing? If it wasn't for endiannes, then this is a bug. If it was for endiannes, then it was likely buggy on a different endiannes machine, and this change is both propagating that and potentially making it worse because of the else block based on the normalizing field.

Reply:
I doubt it is because of endiannes, but I am not sure to be honest.
The [reference source](https://referencesource.microsoft.com/#system.data/system/data/Sql/sqlnorm.cs) seems to list the contributors to the file at the top, maybe ask them, if they still work at MS?
@beysims 

Regardless, _skipNormalize never seems to be set to anything other than false, so the only execution path that is taken is the Big Endian one.

Reply:
> so the only execution path that is taken is the Big Endian one.

In which case the code will end up behaving differently with this change if it's run on a big-endian machine.

Maybe @saurabh500 can comment on this further.

Is this code on a hot path anywhere?  Do you have a perf test that exercises this code in a real use-case and where this ends up showing up?  Unless it's really meaningful, I don't think the benefit is worth the churn/risk.

Reply:
This code path is used while processing Udt (User defined types) from Sql Server which is a CLR DLL deployed on Sql Server. The usage is largely dependent on the client application. 

I don't have a good understanding of this area and cannot validate the correctness of this change. 

If can @bbartels can explain why this change would work correctly, then we could figure out the test coverage, else eliminate changes in this area.

Reply:
My assumption was that BinaryPrimitives will always write in the byte order specified (Big/Little Endian). With that assumption we can use this to our advantage to circumvent the need to reverse the array before writing it to the stream. Now I may very well be wrong with that assumption, so please correct me if I am.

Reply:
> My assumption was that BinaryPrimitives will always write in the byte order specified (Big/Little Endian)

It will.  But BitConverter uses the current endianness, so this is changing from using the opposite of the current endiness to using big endian, which means it's a behavior change if the machine's endianness is big.

Reply:
Ah, I see what you mean now, my bad. Yes that would indeed introduce side-effects on big-endian machines. I suppose if this genuinely has no practical benefits, might as well revert it. Otherwise I can just change it to use the span based BitConverter Api (and manually reverse the span), then it should work just the same as it did before.

Reply:
Switched to BitConverter, the code is now very similar to the previous implementation. 
https://github.com/dotnet/corefx/pull/36838/commits/d9a35e7b1bba194e50e37eb0169ed1cb5474cfb4
Though if you still feel like reverting, I'll do that

Reply:
Also just scoured through the call hierarchy of the NormalizedSerializer (took me too long) and it turns it it is definitely on a hot path. It propagates all the way up to [ExecuteNonQuery](https://apisof.net/catalog/System.Data.SqlClient.SqlCommand.ExecuteNonQuery()).
https://github.com/dotnet/corefx/blob/5458fe268482522aa5681f001157ba7c0a451561/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs#L859-L873

https://github.com/dotnet/corefx/blob/5458fe268482522aa5681f001157ba7c0a451561/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs#L1145-L1176
https://github.com/dotnet/corefx/blob/5458fe268482522aa5681f001157ba7c0a451561/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs#L2371-L2420
Hot-Path here
https://github.com/dotnet/corefx/blob/5458fe268482522aa5681f001157ba7c0a451561/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs#L1574-L1588
(This is in the same function as snipped above)

https://github.com/dotnet/corefx/blob/5458fe268482522aa5681f001157ba7c0a451561/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs#L2066-L2075
https://github.com/dotnet/corefx/blob/5458fe268482522aa5681f001157ba7c0a451561/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs#L3106-L3150
https://github.com/dotnet/corefx/blob/5458fe268482522aa5681f001157ba7c0a451561/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs#L1632-L1656

https://github.com/dotnet/corefx/blob/5458fe268482522aa5681f001157ba7c0a451561/src/System.Data.SqlClient/src/System/Data/Sql/SqlSer.cs#L43

https://github.com/dotnet/corefx/blob/5458fe268482522aa5681f001157ba7c0a451561/src/System.Data.SqlClient/src/System/Data/Sql/SqlSer.cs#L109-L119

So assuming TdsEnums.SQLRETURNVALUE is hit often enough (the name suggests it does) and Format.UserDefined isn't overly prevalent, then I think it most definitely will have a decent impact on reducing allocations. I would try to benchmark this, but I am not sure how I would go about this.
Does dotnet/performance have any benchmarks relating to this?

Reply:
@saurabh500 I did some more investigation to see in which scenarios the code path is executed. From reading the [TDS specification](https://docs.microsoft.com/en-us/openspecs/windows_protocols/MS-TDS/b46a581a-39de-4745-b076-ec4dbb7d13ec) the control token that needs to be returned from the server to activate this code path is defined in section 2.2.7.18(RETURNVALUE). Do you have any insights/guesses on how often such a scenario occurs in production code? It seems rather fringe, in which case it would probably be better to undo this change.

Reply:
As the changes to SqlClient seem to be too risky to offset the potential gain I reverted the changes I made. Closing this discussion.